docs/testing-guide.md

```diff
@@ -0,0 +1,320 @@
+
+### Mocking
+
+- Use mocks and stubs to isolate the unit under test from external dependencies.
+- Use a mocking library like `sinon` to create mocks and stubs.
+- Avoid over-mocking. Only mock what is necessary to isolate the unit under test.
+- Prefer spies to stubs when possible to verify that a method was called without affecting its behavior.
+- Use `sinon-chai` to make assertions on spies and stubs. (e.g., `expect(spy).to.have.been.calledOnce`).
+- Avoid mocking the system under test. Only mock external dependencies.
+
+### Performance
+
+- Ensure that tests run quickly to avoid slowing down the development process.
+- Use parallel test execution where possible to speed up the test suite.
+- Do not use sleep statements or other artificial delays in tests to wait for asynchronous operations to complete.
+  Instead, spy on the asynchronous operation and await it in the test to ensure it has completed.
+- Make sure that any common setup which is expensive is done only once and reused across tests.
+- Do not rely on **_external_** services or databases in unit tests to avoid slowing down the test suite.
+- Use an in-memory test database or mock external services in integration tests to speed up test execution.
+
+### Documentation
+
+- The combined title of the `describe` and `it` blocks should be descriptive enough for the reader to understand what is
+  being tested.
+- Anything which is not self-explanatory from the combined title of the `describe` and `it` should be documented with
+  comments.
+- Use comments to explain WHYs and not WHATs. The WHAT should be clear from the test itself.
+
+## Common Pitfalls
+
+### Flaky Tests
+
+- Ensure tests are deterministic and do not rely on external factors.
+- Use mocks and stubs to isolate tests.
+
+### Overlapping Tests
+
+- Avoid dependencies between tests.
+- Ensure each test can run independently.
+
+## Formatting
+
+### General Rules
+
+- Use consistent formatting for tests to make them easy to read and understand.
+- Use a consistent naming convention for tests to make them easy to identify.
+- Use descriptive variable names to make the test code self-explanatory.
+- Use `before`, `beforeEach`, `after`, and `afterEach` hooks to set up and tear down test fixtures.
+- Use `describe` blocks to group related tests together.
+
+### Naming and Structure
+
+- Use the following naming convention for test files: `<module under test>.spec.ts`
+- The outermost describe should include the name of the module in the test suite:
+  `describe('<name of module/class under test>', () => { ... })`.
+- Multiple tests over the same method should be grouped under the name of the method:
+  `describe('<name of method under test>', () => { ... })`.
+- Use the following naming convention for nested groups of tests:
+  `describe('given <common condition for group of tests>', () => { ... })`.
+- Use the following naming convention for test cases: `it('should <expected behavior of method>', () => { ... })`.
```

Comment:
is there any particular reason to use arrow functions `() => {}` instead of `function() {}`? From https://mochajs.org/#arrow-functions

> Passing [arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) (aka “lambdas”) to Mocha is discouraged. Lambdas lexically bind this and cannot access the Mocha context. For example, the following code will fail:
>
> ```js
> describe('my suite', () => {
>   it('my test', () => {
>     // should set the timeout of this test to 1000 ms; instead will fail
>     this.timeout(1000);
>     assert.ok(true);
>   });
> });
> ```
>
> If you do not need to use Mocha’s context, lambdas should work. Be aware that using lambdas will be more painful to refactor if the need eventually arises!

Reply:
Fair point, I will replace lambdas with `function ()` statements